### PR TITLE
Add JWT parsing functions to ext::auth

### PR DESCRIPTION
**Pull request by elprans**
With this, setting the `ext::auth::client_token` global will yield an
identity via the `ext::auth::ClientTokenIdentity` global if the token is
valid.

### CODE REVIEW

File: edb/lib/ext/auth.edgeql [hunk 1]
```diff
@@ -48,4 +48,114 @@ CREATE EXTENSION PACKAGE auth VERSION '1.0' {
...
+                        false, message := "unsupported JWT algo")
+                ),
+            select
+                std::to_json(
+                    std::to_str(
+                        std::enc::base64_decode(
+                            jwt.payload,
+                            padding := false,
+                            alphabet := std::enc::Base64Alphabet.urlsafe,
+                        ),
```
Comment (scotttrinh) [inline edb/lib/ext/auth.edgeql:77]: This is lovely ❤️‍🔥

File: edb/lib/ext/auth.edgeql [hunk 2]
```diff
@@ -48,4 +48,114 @@ CREATE EXTENSION PACKAGE auth VERSION '1.0' {
...
+                )
+            order by
+                assert(
+                    std::enc::base64_encode(
+                        ext::pgcrypto::hmac(msg, key, hash),
+                        padding := false,
+                        alphabet := std::enc::Base64Alphabet.urlsafe,
+                    ) = jwt.signature,
+                    message := "JWT signature mismatch",
+                )
```
Comment (scotttrinh) [inline edb/lib/ext/auth.edgeql:88]: Out of curiosity, why are the asserts in the `order by` clause?

**Review by scotttrinh — APPROVED** [review]
This is really great! Reads well and easy to understand. 🏆

**Comment by scotttrinh** [comment]
> With this, setting the ext::auth::client_token global will yield an identity via the ext::auth::ClientTokenIdentity global if the token is valid.

What happens when there is no `client_token` set is it just an empty set?

File: edb/lib/ext/auth.edgeql (hunk 2 above)
Comment (raddevon) [reply edb/lib/ext/auth.edgeql:88]: It makes sure the assertion only happens on objects that pass your filter.

File: edb/lib/ext/auth.edgeql [hunk 3]
```diff
@@ -48,4 +48,114 @@ CREATE EXTENSION PACKAGE auth VERSION '1.0' {
...
+                    t := ext::auth::_jwt_check_signature(
+                        ext::auth::_jwt_parse(token),
+                        key,
+                        algo,
+                    ),
+                },
+                validity_range := std::range(
+                    std::to_datetime(<float64>json_get(jwt.t, "nbf")),
+                    std::to_datetime(<float64>json_get(jwt.t, "exp")),
+                ),
```
[the remark this answers is not in the document]
Comment (elprans) [reply edb/lib/ext/auth.edgeql:131]: That's fine, it'll be an unbounded range in that case.  If neither is set, then it's an infinite range which will pass the check.

**Comment by elprans** [comment]
> What happens when there is no `client_token` set is it just an empty set?

Yes

**Comment by elprans** [comment]
The test failure will go away once this is rebased on master.

File: edb/lib/ext/auth.edgeql (hunk 2 above)
Comment (elprans) [reply edb/lib/ext/auth.edgeql:88]: Semantically it makes more sense to put assertions in `filter`, but that affects result cardinality inference, so it's in `order by`.  Assertions need to be placed in a way that makes sure they actually get executed (just putting them in a `with` and not using the alias will not work).

File: tests/test_http_ext_auth.py [hunk 4]
```diff
@@ -171,6 +171,29 @@ class TestHttpExtAuth(tb.ExtAuthTestCase):
...
 
+    @classmethod
+    def get_setup_script(cls):
+        res = super().get_setup_script()
+
+        import os.path
+
+        # HACK: As a debugging cycle hack, when RELOAD is true, we reload the
+        # extension package from the file, so we can test without a bootstrap.
+        RELOAD = True
```
Comment (msullivan) [inline tests/test_http_ext_auth.py:182]: Mark it False

**Comment by msullivan** [comment]
Are we ready to merge this?

**Pull request merged**